### PR TITLE
WIP:  A draft PR for collaborating on ZIP155

### PR DESCRIPTION
**Pull request by zancas**
High Level
========

@AloeareV, @dannasessha and I began this project by inspecting the PR that merged TorV3 addresses into bitcoin `main`.

From the perspective of "bitcoin main" circa 2019, there are two upgrades that were required. Add support for BIP155-style version 2 network addresses, part of [BIP155](https://github.com/bitcoin/bips/blob/master/bip-0155.mediawiki):

1.  https://github.com/bitcoin/bitcoin/pull/19031

And, given support for BIP155 `addrv2`s, add support for TorV3 addresses:

2.  https://github.com/bitcoin/bitcoin/pull/19954

Adding Support for BIP155
====================

The addition of `addrv2` support was initiated by adding regression tests:

3.  https://github.com/bitcoin/bitcoin/pull/19351

When we inspected that regression test code, we saw that the `netbase` code had been updated (relative to zcash `main` circa 2021) to reduce the responsibilities of the `CNetAddr`, `CService`, and `CSubNet` classes.  In particular newer versions are not responsible for name resolution.

We discovered that we minimally required:

  https://github.com/bitcoin/bitcoin/pull/8128/commits/b6c3ff3daed432e71d1c5bfe2357c4bb9fdc862a

In order to proceed with the backport of 19351.

We noted the coherent structure of:

4.  https://github.com/bitcoin/bitcoin/pull/8128

We surmised that all of 8128 would eventually be relevant for TorV3 support, and further noted it's dependence on:

5.  https://github.com/bitcoin/bitcoin/pull/7868

Following the above reasoning we backported 7868 which depended on a small upgrade to error message formatting:

6. https://github.com/bitcoin/bitcoin/pull/7257

The 

We need this code for `NODE_NONE`:

https://github.com/bitcoin/bitcoin/pull/7749/commits/ee06e04369c37da21e048fda849cce2a1f066f84

We cherry-picked https://github.com/bitcoin/bitcoin/pull/13575/commits/07c493f2d1053d6a1c7d0ca1650da7b6f94c8948 for the `NET_TOR` to `NET_ONION` change.

This commit introduces GetAddrBytes:   https://github.com/bitcoin/bitcoin/pull/19219/commits/b691f2df5f7d443c0c9ee056ab94aa0fc19566d5

This commit introduces the `NET_INTERNAL` network variant:   https://github.com/bitcoin/bitcoin/pull/10446/commits/7f31762cb6261806542cc6d1188ca07db98a6950

The FLEXIBLE_SIZE change is part of prepation for us:

https://github.com/bitcoin/bitcoin/pull/19628/commits/1ea57ad67406b3aaaef5254bc2fa7e4134f3a6df

We're going to need to backport span, prior to other bps (Making `CNetAddr` members variable length), because `Span` is used in core methods.

ADDING SUPPORT FOR BITCOIN STYLE SERIALIZATION:

https://github.com/bitcoin/bitcoin/pulls?q=is%3Apr+is%3Amerged+%22Serialization+improvements%22+in%3Atitle

NOTES:
  *  I use `main` as slang for `master` where I deem it to be (sufficiently) unambiguous.

### CODE REVIEW

File: src/net.cpp [hunk 1]
```diff
@@ -144,7 +144,7 @@ static std::vector<CAddress> convertSeed6(const std::vector<SeedSpec6> &vSeedsIn
     {
         struct in6_addr ip;
         memcpy(&ip, i->addr, sizeof(ip));
-        CAddress addr(CService(ip, i->port));
+        CAddress addr(CService(ip, i->port), NODE_NETWORK);
```
Comment (zancas) [inline src/net.cpp:147]: @AloeareV I am still skeptical about the definition of `NODE_NETWORK`:

`NODE_NETWORK = (1 << 0)`

but your hypothesis that it's to make it explicit that the value is to be used as a bit-flag is consistent with other patterns we've seen in C++ code (e.g. type hints in names), so it's plausible.

File: src/net.cpp [hunk 2]
```diff
@@ -1636,7 +1638,12 @@ void ThreadOpenAddedConnections()
         for (vector<CService>& vserv : lservAddressesToAdd)
         {
             CSemaphoreGrant grant(*semOutbound);
-            OpenNetworkConnection(CAddress(vserv[i % vserv.size()]), &grant);
+            OpenNetworkConnection(CAddress(vserv[i % vserv.size()], 0), &grant);
+            //=======
+            ///* We want -addnode to work even for nodes that don't provide all
+            // * wanted services, so pass in nServices=0 to CAddress. */
+            //OpenNetworkConnection(CAddress(vserv[i % vserv.size()], 0), false, &grant);
+            //>>>>>>> 15bf863219 (Don't require services in -addnode)
```
Comment (AloeareV) [inline src/net.cpp:1646]: We can probably remove the commented diff, the bool in the commented arg is only something other than false in bitcoin in a single call in `ThreadOpenConnections`, which looks to have been significantly refactored in zcash.

File: src/net.cpp (hunk 1 above)
Comment (zancas) [reply src/net.cpp:147]: https://www.stroustrup.com/C++11FAQ.html <-- Let's stay aware of the `strong enum` pattern.

File: src/protocol.h [hunk 3]
```diff
@@ -87,7 +87,7 @@ class CAddress : public CService
 {
 public:
     CAddress();
-    explicit CAddress(CService ipIn, uint64_t nServicesIn = NODE_NETWORK);
+    explicit CAddress(CService ipIn, uint64_t nServicesIn);
```
Comment (zancas) [inline src/protocol.h:90]: IMHO this move toward explicitness is healthy.

File: src/protocol.cpp [hunk 4]
```diff
@@ -79,7 +79,7 @@ CAddress::CAddress(CService ipIn, uint64_t nServicesIn) : CService(ipIn)
 
 void CAddress::Init()
 {
-    nServices = NODE_NETWORK;
+    nServices = 0;
```
Comment (zancas) [inline src/protocol.cpp:82]: +1

File: src/net.cpp (hunk 2 above)
Comment (zancas) [reply src/net.cpp:1646]: Alright let's leave this comment thread open for now.  Likely we'll circle back and clean up, per your recommendation.

**Comment by zancas** [comment]
I am reorganizing this code, will track elsewhere.

**Pull request closed without merging**